### PR TITLE
Allow IsoSecond to grow up to 70

### PR DESCRIPTION
**Pull request by Ekleog**
In order to represent times just before Jan 1, 1972, we need a number of second that grows up enough to accomodate for the 10 leap seconds that occurred (~assuming the IERS definition of UTC; if [hifitime is correct](https://docs.rs/hifitime/latest/src/hifitime/leap_seconds.rs.html#44-58) then SOFA claimed non-integer leap seconds, but even with this definition of UTC we would still need more than one at a time as they still had a roughly 5s jump at the posix epoch~ edit: this is wrong and actually more complex, see https://github.com/unicode-org/icu4x/pull/3087#issuecomment-1420267712 and https://github.com/unicode-org/icu4x/pull/3087#issuecomment-1425058858).

In addition to changing the constant, I also explained in the doc-comments what I learned on https://github.com/unicode-org/icu4x/discussions/3086 :)

(Note to self: submit as a trophy case for cargo-bolero if this is confirmed to be amiss, as my fuzzer for kine found it)

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

**Comment by zbraniecki** [comment]
> In order to represent times just before the POSIX epoch

Excuse my ignorance but why would we want to do this?

If I understand this will not affect the size of the type itself, but will require one more bit allocation which may affect compact storing like we use in ZeroVec.

So, there's a cost, and I'm not sure if I understand the value.

**Comment by eggrobin** [comment]
> SOFA claimed non-integer leap seconds

Pre-1972 UTC had non-integer leaps, but it also had changes in _rate_; the seconds should still be in their normal range.
See, e.g., https://hpiers.obspm.fr/iers/bul/bulc/UTC-TAI.history.

**Review by Manishearth — CHANGES_REQUESTED** [review]
Going to cancel my approving review due to Robin's comment

**Comment by Ekleog** [comment]
@eggrobin Interesting, thank you! I hadn't known of that subtlety of this leap second accounting scheme :) My bad for not having checked my sources on which leap seconds the IERS claims happened.

That said, my current understanding is they're relying on a definition of UT"C" which was not necessarily well-acknowledged in time. For instance, [NIST does claim there was a 10 second leap](https://www.nist.gov/pml/time-and-frequency-division/leap-seconds-faqs).

Arguably, UTC did not actually exist before 1972. [The IETF](https://www.ietf.org/timezones/data/leap-seconds.list) confirms this view, with the following paragraph:
```
#	3. The current definition of the relationship between UTC
#	and TAI dates from 1 January 1972. A number of different
#	time scales were in use before that epoch, and it can be
#	quite difficult to compute precise timestamps and time
#	intervals in those "prehistoric" days. For more information,
#	consult: [...]
```

Unfortunately, the "scale the time before 1972" approach does not actually yield itself to being written down with the "hours, minutes, seconds, nanoseconds" scheme currently employed by icu4x, which can only support leap seconds located at the last second of a minute. So I think only the NIST definition of UTC would be usable in the context of presenting the user with information they can actually act on, though alternative implementations using the rescaled time and eg. showing leap seconds at the end of the minute where they actually happened would make sense too.

Now, I would totally understand if performance concerns meant you don't actually want to support this use case which is, granted, very niche. (Though I'm not sure what kind of code would ever have ICU date handling as a hot code path)
If you don't think it's useful, I'll just make conversion to ICU time stall for 10 seconds in 1972 when in an NIST-based timezone and that will certainly be fine :)

**Comment by Ekleog** [comment]
I've just pushed fixed comments about what exactly the 70 limit is about, should you want to merge this after weighing the pros and cons :)

Feel free to close if the answer is no, I'd totally understand!

**Comment by eggrobin** [comment]
Apologies for the delayed response, I was on vacation.

TL;DR: the proposed change is based on a misunderstanding of the relevant time standards; no time stamps of the form that this PR proposes supporting should be expected to have been recorded nor interchanged.

### 1. On UTC.
> That said, my current understanding is they're relying on a definition of UT"C" which was not necessarily well-acknowledged in time.

> [...]

> Arguably, UTC did not actually exist before 1972.

Coordinated Universal Time (back then a standard for radio time signal) was defined by recommendation 374-1 of the CCIR of the ITU-R at its XIth plenary assembly [CCIR1966] (in continuity with the definitions of recommendation 374 at its Xth plenary assembly [CCIR1963] and recommendation 319 at its IXth plenary assembly [CCIR1959]), and is referenced by the IAU in resolution 3 of its XIIIth General Assembly (1967). There was no ambiguity as to what it was, and its frequency offsets (changes in rate) and epoch adjustments (leaps) were coordinated internationally and across the various standardization bodies.

In particular, the time signals transmitted by NIST’s WWV and WWVH were among those coordinated in this manner; see Section 1.5(b) in [NIST1965].

No scare quotes are needed around the C. There are other things whose name starts with UT (notably UT1 and UT2, see below), but there is only one UTC.

#### References
- [CCIR1959] https://search.itu.int/history/HistoryDigitalCollectionDocLibrary/4.275.43.en.1001.pdf
- [CCIR1963] https://search.itu.int/history/HistoryDigitalCollectionDocLibrary/4.276.43.en.1003.pdf
- [CCIR1966] https://search.itu.int/history/HistoryDigitalCollectionDocLibrary/4.277.43.en.1003.pdf
- [IAU1967] https://www.iau.org/static/resolutions/IAU1967_French.pdf
- [NIST1965] https://tf.nist.gov/general/pdf/1718.pdf

---

> [The IETF](https://www.ietf.org/timezones/data/leap-seconds.list) confirms this view, with the following paragraph:

I contend that this is a misreading of that comment in the IETF leap seconds file.

> ```
> #	3. The current definition of the relationship between UTC
> #	and TAI dates from 1 January 1972.
> ```

This is true; the current system of integral leaps at the end of June or December can be traced back to recommendation 460 of the CCIR at its XIIth plenary assembly [CCIR1970], and the resolution adopted by Commission 31 of the IAU at its XIVth General Assembly [IAU1970]. These take effect on the 1st of January 1972 at 00:00 UTC (note that the definition has since been amended multiple times; the latest version is [ITU-R-TF.460-6], 2002).
The leap at 1972-01-01T00:00:00Z is a fraction of a second, see 4. in the IAU resolution (from the aforementioned IERS file, the offset is readily calculated as 0.107758 s, a number which may easily be corroborated, _e.g._, against Table 9 of [BIH1985]).

- Note: In general, while the definitions of time scales change every now and then as metrology improves, the new definition preserves continuity with the old one (see for instance Ephemeris Time, defined by the resolution of commissions 4 and 31 of [IAU1958], being replaced by TT in Recommendation IV of Resolution A4 of [IAU1991], and TT being redefined as a linear scaling of TCG by resolution B1.9 of [IAU2000]).

#### References
- [BIH1985] https://webtai.bipm.org/ftp/pub/tai/annual-reports/bih-annual-report/BIH_1985.pdf
- [CCIR1970] https://search.itu.int/history/HistoryDigitalCollectionDocLibrary/4.278.43.en.1004.pdf
- [IAU1958] https://www.iau.org/static/resolutions/IAU1958_French.pdf
- [IAU1970] https://www.iau.org/static/resolutions/IAU1970_French.pdf
- [IAU1991] https://www.iau.org/static/resolutions/IAU1991_French.pdf
- [IAU2000] https://www.iau.org/static/resolutions/IAU2000_French.pdf
- [ITU-R-TF.460-6] https://www.itu.int/rec/R-REC-TF.460-6-200202-I/en

> ```
>                                       A number of different
> #	time scales were in use before that epoch,
> ```

This is also true; notably UT2 was a commonly-used flavour of universal time, which UTC was trying to approximate (UTC now tries to approximate UT1, whose current definition is based on the Earth Rotation Angle; see Section 5.3.3, _Earth Rotation Angle_, in the IERS Conventions (2010) [IERS-TN36]).
Things are also a complete mess before time signals started being internationally coordinated in 1961.

#### References
- [IERS-TN36] https://www.iers.org/SharedDocs/Publikationen/EN/IERS/Publications/tn/TechnNote36/tn36.pdf

> ```
>                                              and it can be
> #	quite difficult to compute precise timestamps and time
> #	intervals in those "prehistoric" days.
> ```

Indeed. Not that most applications would care, but the length of this discussion should make that plain.

### 2. On the idea of a 10-second leap.

> For instance, [NIST does claim there was a 10 second leap](https://www.nist.gov/pml/time-and-frequency-division/leap-seconds-faqs).

> […] the NIST definition of UTC […]

This page is not a definition, nor anything normative; it is a high-level FAQ, describing the current state of things (that is, UTC as defined by CCIR recommendation 460) rather than the historical arcana. The ten-second offset refers to the difference between TAI and UTC when CCIR recommendation 460 takes effect.

This does not mean that anyone switched from TAI to UTC with a ten-second leap during which they would produce the timestamps that this PR purports to support; since TAI is not interchanged, anyone who was using TAI was likely one of the national laboratories contributing to it, and still uses it; anyone getting their time signals by radio (for instance from NIST’s WWV and WWVH) would have been using the variable-frequency UTC, and would have switched to the current scheme of UTC—transmitted by the same signals—with the 0.107758 s leap mentioned above.

In other words: _no-one producing time stamps according to any standards in effect at the time would have recorded a time stamp of 1971-12-31T23:59:69Z_, and applications handling standard time stamps should never have to handle such things.

### 3. On practical concerns.

> Unfortunately, the "scale the time before 1972" approach does not actually yield itself to being written down with the "hours, minutes, seconds, nanoseconds" scheme currently employed by icu4x, which can only support leap seconds located at the last second of a minute.

The adjustments to the epoch (leaps) were made only at 0h UTC on the first day of a month (see CCIR recommendation 374-1, 4.3 and 4.4), so those UTC time stamps are representable using the normal scheme.

> […] though alternative implementations using the rescaled time and eg. showing leap seconds at the end of the minute where they actually happened would make sense too.

The scaling is of no concern unless you are converting timestamps before 1972 between TAI and UTC (in which case you are a national laboratory), or between TT and UTC (in which case you might be an astronomer rather than a metrologist).

That is, the scaling is a problem only if you need to relate SI seconds at the geoid, as measured with the precision of an atomic clock, to time intervals of UTC, between 1961 and 1972.
Otherwise, it is irrelevant that the time between 1970-01-01T00:00:00Z and 1970-01-01T00:00:01Z, as measured by the atomic clocks defining TAI, is 1.000 000 030 s; you just use seconds of UTC for your durations, rather than seconds of TT(TAI).

Although conversion between the various realizations of TT (or TCG) is an interesting problem, I do not believe ICU4X is meant as a metrological nor an astronomical timescale conversion library.

> So I think only [the nonstandard time scale obtained by gluing TAI to UTC at 1972-01-01T00:00:00Z and sticking ten intercalary seconds there] would be usable in the context of presenting the user with information they can actually act on,

If a user cares about this level of precision for old time stamps, silently showing that user TAI before 1972 with an _ad hoc_ leap second scheme seems unlikely to be a good idea.

In any case, ICU4X-SC should not be in the business of defining new time standards; we have enough of those already, in the hands of enough committees and commissions.

File: components/calendar/src/types.rs [hunk 1]
```diff
@@ -293,21 +293,32 @@ dt_unit!(
...
 
 dt_unit!(
     IsoMinute,
     u8,
     60,
-    "An ISO-8601 minute component, for use with ISO calendars.\n\nMust be within inclusive bounds `[0, 60]`."
+    "An ISO-8601 minute component, for use with ISO calendars.
+
+Must be within inclusive bounds `[0, 60]`. The value could be equal to 60 to
+indicate the end of an hour, with the writing 12:60:00."
```
Comment (eggrobin) [inline components/calendar/src/types.rs:309]: Like the note above for hours, this clarification of the semantics is a good idea.

Contrary to the hours above (for which the meaning of 24 is defined by Section 4.2.3 in ISO 8601:2004), this is an extension to ISO 8601; this should be noted, and a similar comment should be made for seconds (with 70 being changed back to 61, see my other comment).

**Comment by Manishearth** [comment]
I'm going to close this PR for now, y'all can continue discussion here and reopen if it turns out there are changes that we should merge.

**Comment by Ekleog** [comment]
@eggrobin Thank you very much for your detailed answer! I'm sorry I misinterpreted anecdotal data and made you have to dig actually normative references, but thank you very much for actually providing me with the correct information! I'll make sure to fix my code to match the actual UTC definition thanks to you :)

@Manishearth I have now updated the PR (and commit message) to only contain the documentation clarification and no actual change to IsoSecond. Would it be possible for you to reopen and review?

BTW, now I've written the full documentation, I'm unsure making this extension to ISO 8601 was a good choice, especially for seconds, as it means that 23:59:60Z could correspond to either the magic "end of minute" marker or half-a-second before 23:59:60.5Z depending on whether there was a leap second, and so it's hard to use the result… but it'd probably be a breaking change and is most likely not actually a problem. (Though I'm not sure why the extension to ISO 8601 was used for minutes and seconds, but not nanoseconds with the same logic)

Anyway, I'm not here to litigate this, only to document the current behavior, I'll let you judge whether you want to change the actual behavior :)

**Comment by Ekleog** [comment]
Well, seems like github isn't catching the update, so let's keep this closed and I've just opened https://github.com/unicode-org/icu4x/pull/3156 with only the documentation changes. Sorry all for taking your time and thank you for the detailed walkthroughs!

**Comment by Manishearth** [comment]
> as it means that 23:59:60Z could correspond to either the magic "end of minute" marker or half-a-second before 23:59:60.5Z depending on whether there was a leap second

I don't think the end-of-minute marker is "magic", there's no special logic about it, it just enables representing that which is convenient in some algorithms. In general being able to go "one over" is useful for this kind of thing. (The C memory model does something similar with pointer validity)
